### PR TITLE
Restructure of CLI and application state

### LINKED ISSUE #2169: Production and development modes (default workers)
While it is possible to specify these individual features by hand, I would rather see a simple mode flag to get sensible defaults:
- Production: no access logs, workers=os.cpu_count()
- Development: debug, access logs, auto-reloader

Sanic server by default runs with a single worker process and doesn't enable features that would be useful in development. Many people accidentally use these default settings for benchmarking and production but the performance is poor due to lack of multiprocessing and due to logging if `app.run` was used. The debug mode cannot be enabled by default because running it in production is a potential security risk, so making the production mode the new default seems a better option.

Notably the reloader setting by default already follows the debug flag. I believe this should done with the access log setting too (currently `app.run(debug=True)` and `app.run(debug=False)` both enable access logs, while `sanic` CLI disables logs irregardless of the `--debug` flag). Alike, the number of workers could depend on the debug flag, keeping the single worker default in debug mode, but using the number of CPUs by default.

By these changes we get **production mode by default**, and adding `debug=True` or `sanic --debug` would accomplish full switch of all relevant settings into development mode.

This is potentially a breaking change with application servers that run multiple Sanic instances on different ports with the intention that each Sanic instance stays a single process. If this is a problem, specifying workers=1 in these installs is a simple and backwards-compatible fix.

Relevant `app.run` parts shown with changes:
```python
    def run(
        debug: bool = False,
        auto_reload: Optional[bool] = None,  # already follows debug flag if not set
        workers: Optional[int] = None,   # instead of: workers: int = 1
        access_log: Optional[bool] = None,  # was already this but doesn't work as expected
    ) -> None:
        # New logic
        if workers is None:
            workers = 1 if debug else os.cpu_count()
        if access_log is None:
            access_log = debug
```

I'd like to hear your thoughts before implementing this. This can probably be done better, or perhaps I overlooked something.

### PR DESCRIPTION
**Pull request by ahopkins**
Closes #2169 
Replaces #2237 

---

## Purpose

This PR started out as a way to add some additional running defaults via the CLI. It has grown in scope a little and is taking on some refactoring of the `Sanic` class (still much more is needed).

The big change (although we strive for backwards compat here, so please LMK if you see anything that is breaking) is to move some properties from the `Sanic` instance to a new `ApplicationState` instance. After v21.12, there is a lot more refactoring to be done.

## Implementation

| Flag    | Mode  | Tracebacks | Logging | Access logs | Reload | Max workers |
|---------|-------|------------|---------|-------------|--------|-------------|
| --debug | DEBUG | yes        | DEBUG   | yes         | ^1     |             |
|         | PROD  | no         | INFO ^2 | ^3          |        |             |
| --dev   | DEBUG | yes        | DEBUG   | yes         | yes    |             |
| --fast  |       |            |         |             |        | yes         |

- ^1 `--debug` to deprecate auto-reloading and remove in 22.3
- ^2 After 22.3 this moves to WARNING
- ^3 After 22.3: no

## Additions

The new features are:

- `--fast`: default to the maximum allowed workers
- `--verbosity`: allows us to hide some logging noise
- Pretty MOTD display

image

## Cleanup

There is also a bit of refactoring of `__main__` to move it into a module. The single function was getting a bit messy to maintain. While I did not touch the bulk of the logic (saved for another time), it does make it a little easier to handle the addition of arguments. And, a side benefit is that they are grouped together a little more nicely.

```
$ sanic -h
usage: sanic [-h] [--version] [--factory] [-s] [-H HOST] [-p PORT] [-u UNIX] [--cert CERT] [--key KEY] [--tls DIR] [--tls-strict-host]
             [-w WORKERS | --fast] [--access-logs | --no-access-logs] [--debug] [-d] [-r] [-R PATH] [--motd | --no-motd] [-v]
             [--noisy-exceptions | --no-noisy-exceptions]
             module

   ▄███ █████ ██      ▄█▄      ██       █   █   ▄██████████
  ██                 █   █     █ ██     █   █  ██
   ▀███████ ███▄    ▀     █    █   ██   ▄   █  ██
               ██  █████████   █     ██ █   █  ▄▄
  ████ ████████▀  █         █  █       ██   █   ▀██ ███████

 To start running a Sanic application, provide a path to the module, where
 app is a Sanic() instance:

     $ sanic path.to.server:app

 Or, a path to a callable that returns a Sanic() instance:

     $ sanic path.to.factory:create_app --factory

 Or, a path to a directory to run as a simple HTTP server:

     $ sanic ./path/to/static --simple

Required
========
  Positional:
    module                         Path to your Sanic app. Example: path.to.server:app
                                   If running a Simple Server, path to directory to serve. Example: ./

Optional
========
  General:
    -h, --help                     show this help message and exit
    --version                      show program's version number and exit

  Application:
    --factory                      Treat app as an application factory, i.e. a () -> <Sanic app> callable
    -s, --simple                   Run Sanic as a Simple Server, and serve the contents of a directory
                                   (module arg should be a path)

  Socket binding:
    -H HOST, --host HOST           Host address [default 127.0.0.1]
    -p PORT, --port PORT           Port to serve on [default 8000]
    -u UNIX, --unix UNIX           location of unix socket

  TLS certificate:
    --cert CERT                    Location of fullchain.pem, bundle.crt or equivalent
    --key KEY                      Location of privkey.pem or equivalent .key file
    --tls DIR                      TLS certificate folder with fullchain.pem and privkey.pem
                                   May be specified multiple times to choose multiple certificates
    --tls-strict-host              Only allow clients that send an SNI matching server certs

  Worker:
    -w WORKERS, --workers WORKERS  Number of worker processes [default 1]
    --fast                         Set the number of workers to max allowed
    --access-logs                  Display access logs
    --no-access-logs               No display access logs

  Development:
    --debug                        Run the server in debug mode
    -d, --dev                      Currently is an alias for --debug. But starting in v22.3, 
                                   --debug will no longer automatically trigger auto_restart. 
                                   However, --dev will continue, effectively making it the 
                                   same as debug + auto_reload.
    -r, --reload, --auto-reload    Watch source directory for file changes and reload on changes
    -R PATH, --reload-dir PATH     Extra directories to watch and reload on changes

  Output:
    --motd                         Show the startup display
    --no-motd                      No show the startup display
    -v, --verbosity                Control logging noise, eg. -vv or --verbosity=2 [default 0]
    --noisy-exceptions             Output stack traces for all exceptions
    --no-noisy-exceptions          No output stack traces for all exceptions
```

### CODE REVIEW

**Comment by ahopkins** [comment]
@Tronic We discussed making `--dev` be `INFO` and not `DEBUG`. That does not really make sense with this implementation since (_eventually_) `--dev` and `--debug` will mean something different: `--dev` will be `--debug` + `--auto-reload`.

To compensate for that, I buried some of the noise with `--verbosity`.

**Comment by Tronic** [comment]
Could you make the mode line display: `mode: production, single worker` or `mode: production, goin' fast with {N} workers` or something alike, avoiding the separate workers line?

Also, the graphic will look bad on MacOS but I guess there is not much to be done about that.

**Comment by Tronic** [comment]
And actually, replace the host and port lines with URLs and avoid the separate goin' fast message we have been printing.

**Comment by ahopkins** [comment]
@Tronic I was planning on changing the host/port like like you suggested because I knew I needed to support unix socket alternative. I agree about the multiple IPs, but that is a much larger PR since we really should think about how to have 1+ Sanic instances to 1+ bindings.

I incorporated your idea about bringing workers into the `mode` line. I think auto-reload should still be its own because I want to also enumerate and additional directories that are added.  Perhaps the line is just omitted if there is no auto-reload.

**Comment by ahopkins** [comment]
A couple more additions:
- Displays versions of some Sanic libs if they are installed in the environment
- Includes an API for adding arbitrary data to MOTD (Might be helpful for both app and plugin developers)

image

**Comment by Tronic** [comment]
Great! Now just drop one of the Goin' fast so that it is not needlessly repeated :)

I did not yet look at the verbosity changes but maybe still make (some of) these messages INFO rather than DEBUG?

**Comment by ahopkins** [comment]
> Great! Now just drop one of the Goin' fast so that it is not needlessly repeated :)

The "goin' fast" in the `mode` only shows up when you are using `--fast`. Also, I consolidated the packages to a single item:

```
packages: sanic-routing==0.7.1, sanic-testing==0.7.0b2, 
          sanic-ext==21.9.0                             
```

> I did not yet look at the verbosity changes but maybe still make (some of) these messages INFO rather than DEBUG?

I am assuming you mean the non-TTY version. I am not sure about how best to handle this. On the one hand you are likely in a prod container. On the other, it is still nice to have the output in your logs. Maybe I'll leave these as INFO for now and we can think about that at a later point.

For now, the verbosity changes I made only touch the signals. But there probable are some other points we could use it.

**Comment by ahopkins** [comment]
FYI - I confirmed on my Windows laptop that the color looks correct using:
- PowerShell (w/ [help of this](https://github.com/sanic-org/sanic/pull/2295/files#diff-6f2e26f3fd8f7306dd217c48a70cccec6c59d9dd4e21bd10fe8862a86fea49c7R13))
- Terminus emulator
- WSL

I will test on OSX later today

File: tests/test_exceptions_handler.py [hunk 1]
```diff
@@ -220,7 +221,11 @@ def lookup(self, exception):
...
 
-    assert caplog.records[0].message == (
+    for record in caplog.records:
+        if record.message.startswith("You are"):
+            break
+
+    assert record.message == (
         "You are using a deprecated error handler. The lookup method should "
         "accept two positional parameters: (exception, route_name: "
         "Optional[str]). Until you upgrade your ErrorHandler.lookup, "
```
Comment (prryplatypus) [inline tests/test_exceptions_handler.py:231]: Wouldn't this assertion have to be nested inside the `for` loop? Or does exiting the loop leave `record` defined as the last iterated item?

**Review by prryplatypus — CHANGES_REQUESTED** [review]

**Comment by prryplatypus** [comment]
Also, seem to be missing tests for the warning when setting `config.LOGO`.

File: sanic/app.py [hunk 2]
```diff
@@ -1427,6 +1457,114 @@ def update_config(self, config: Union[bytes, str, dict, Any]):
...
+
+    @is_stopping.setter
+    def is_stopping(self, value: bool):
+        self.state.is_stopping = value
+
+    @property
+    def reload_dirs(self):
+        return self.state.reload_dirs
+
+    def motd(self, serve_location):
```
[the remark this answers is not in the document]
Comment (ahopkins) [reply sanic/app.py:1509]: It is a Linuxism: https://linuxhint.com/show-motd-in-linux/

File: sanic/app.py [hunk 3]
```diff
@@ -159,45 +176,42 @@ def __init__(
...
+        self.config: Config = config or Config(
             load_env=load_env, env_prefix=env_prefix
         )
-        self.configure_logging = configure_logging
-        self.ctx = ctx or SimpleNamespace()
-        self.debug = None
-        self.error_handler = error_handler or ErrorHandler(
+        self.configure_logging: bool = configure_logging
+        self.ctx: Any = ctx or SimpleNamespace()
+        self.debug = False
```
[the remark this answers is not in the document]
Comment (ahopkins) [reply sanic/app.py:200]: Backwards compat :disappointed: 

One of my top priorities for v22 is to clean up `app.py`. This includes not only actually breaking the module up into multiple files to make it easier to work with, but also to handle some inconsistencies like that. There are a bunch of properties that I want to move out. The `ApplicationState` is a major first step towards that goal. I was debating whether to actually add these two values to it. I think ideally the `Sanic` object has only the containers on it (router, middleware, etc). Ultimately I decided that `debug` and `configure_logging` are not really state, and (as you mentioned) more appropriately placed on the `Config` object.

However, this PR has enough going on that messing around with the `Config` object seemed overkill and out of scope. I am not opposed to moving as long as there are some setters like I did here for the `ApplicationState` properties. But, I am also not opposed to just waiting until January to move those.

File: sanic/app.py [hunk 4]
```diff
@@ -1033,12 +1056,23 @@ def run(
...
+        }.items():
+            if value is not None:
+                setattr(self.config, attribute, value)
+
+        if fast:
+            self.state.fast = True
+            try:
+                workers = len(os.sched_getaffinity(0))
+            except AttributeError:
+                workers = os.cpu_count() or 1
```
[the remark this answers is not in the document]
Comment (ahopkins) [reply sanic/app.py:1075]: :100: Thank you for catching this. Everything from the CLI should function as if using `run()` and vice versa.

File: tests/test_exceptions_handler.py (hunk 1 above)
Comment (ahopkins) [reply tests/test_exceptions_handler.py:231]: Correct. It is not scoped. `record` is still set from the last iteration of the loop when it was broken.

**Pull request merged**